File: executable-operator/executable-flink/src/main/java/fdu/daslab/executable/flink/operators/FileSink.java
package fdu.daslab.executable.flink.operators;

import fdu.daslab.executable.basic.model.OperatorBase;
import fdu.daslab.executable.basic.model.ParamsModel;
import fdu.daslab.executable.basic.model.ResultModel;
import org.apache.commons.lang.StringUtils;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.thrift.TException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;

/**
 * Flink批数据写入文件的算子
 * @author 李姜辛
 * @version 1.0
 * @since 2021/3/14 15:52
 */
public class FileSink  extends OperatorBase<DataSet<List<String>>, DataSet<List<String>>> {

    Logger logger = LoggerFactory.getLogger(FileSource.class);

    public FileSink(String id, List<String> inputKeys, List<String> outputKeys, Map<String, String> params) {
        super("FlinkFileSink", id, inputKeys, outputKeys, params);
    }


//    // 是否输出一个文件
//    @Parameter(names = {"--isCombined"})
//    Boolean isCombined = Boolean.TRUE;

    @Override
    public void execute(ParamsModel inputArgs,
                        ResultModel<DataSet<List<String>>> result) {
        // FileSink fileSinkArgs = (FileSink) inputArgs.getOperatorParam();
        // 写入文件

        final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
        String separator = this.params.get("separator");
        String outputPath = this.params.get("outputPath");

        final DataSet<String> res = this.getInputData("data")
                .map((MapFunction<List<String>, String>) line -> (String) StringUtils.join(line, separator));


        try {
            final List<String> collect = res.collect();
            Files.write(Paths.get(outputPath), collect);
        } catch (Exception e) {
            e.printStackTrace();
        }


        // TODO: 设置并行度

//        boolean isCombined = this.params.get("isCombined").equals("true"); // todo 之后会根据数据类型在外面自动转换
//        if (isCombined) {
//            // 为了方便其他的节点交互，提供将所有Partition写入一个文件的可能性
//            this.getInputData("data")
//                    .foreachPartition(partitionIter -> {
//                        // 所有数据均追加到一个文件上
//                        FileWriter fileWritter = new FileWriter(this.params.get("outputPath"), true);
//                        PrintWriter out = new PrintWriter(fileWritter);
//                        partitionIter.forEachRemaining(record -> {
//                            StringBuilder writeLine = new StringBuilder();
//                            record.forEach(field -> {
//                                writeLine.append(field);
//                                writeLine.append(this.params.get("separator"));
//                            });
//                            writeLine.deleteCharAt(writeLine.length() - 1);
//                            out.println(writeLine);
//                            out.flush();
//                        });
//                        out.close();
//                        fileWritter.close();
//                    });
//        } else {
//            // 一个partition写入一个文件
//            this.getInputData("data")
//                    .map(line -> StringUtils.join(line, this.params.get("separator")))
//                    .saveAsTextFile(this.params.get("outputPath"));
//        }
        try {
            // 数据准备好
            this.getMasterClient().postDataPrepared();
        } catch (TException e) {
            e.printStackTrace();
        }
    }

}

File: executable-operator/executable-sparkml/main/sparkml/operators/FileSink.py
import traceback

from executable.basic.model.OperatorBase import OperatorBase
from executable.basic.utils.Logger import Logger

"""
@ProjectName: CLIC
@Time       : 2020/12/21 18:50
@Author     : Jimmy
@Description: 将Dataframe存到csv文件中，可选择直接输出（即输出Spark分布式文件）或者转换成 pandas 的 dataframe 后输出单一文件
"""

logger = Logger('OperatorLogger').logger


class FileSink(OperatorBase):
    def __init__(self, ID, inputKeys, outputKeys, Params):
        super().__init__("FileSink", ID, inputKeys, outputKeys, Params)

    def execute(self):
        try:
            data = self.getInputData("data")
            header = self.params["header"].lower()  # 是否保留列名
            if self.params["outputType"] == "spark":
                data.write.option("header", header).csv(self.params["outputPath"])
            elif self.params["outputType"] == "pandas":
                data.toPandas() \
                    .to_csv(self.params["outputPath"], header=(header == "true"), encoding='utf_8_sig', index=False)

            self.setOutputData("result", "Successful!")

        except Exception as e:
            logger.error(e.args)
            logger.error(traceback.format_exc())
